cirq/linalg/transformations.py

```diff
@@ -304,3 +307,109 @@ def partial_trace(tensor: np.ndarray,
     left_indices = [keep_map[i] if i in keep_set else i for i in range(ndim)]
     right_indices = [ndim + i if i in keep_set else i for i in left_indices]
     return np.einsum(tensor, left_indices + right_indices)
+
+
+def wavefunction_partial_trace(wavefunction: np.ndarray,
+                               keep_indices: List[int],
+                               atol: Union[int, float] = 1e-8) -> Tuple:
+    """Computes a mixture representing the partial trace of a wavefunction.
+
+    The input wavfunction must have shape `(2,2,2,...)`. Attempt to factor the
+    wavefunction into a pure state over `keep_indices`. If this fails, a non-
```

Comment:
I would be a bit vaguer here and just say that this method is more efficient for pure states, and describe that running time (e.g. it's `O(2^n)` where n is the number of qubits, whereas for mixed states it will I think be `O(2^n + 4^k)` or maybe even `+8^k` where k is the number of kept qubits.)